plugins/events-backend-module-http/src/service/HttpEndpointEventPublisher.ts

```diff
@@ -0,0 +1,101 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { errorHandler } from '@backstage/backend-common';
+import { Config } from '@backstage/config';
+import { EventPublisher, EventBus } from '@backstage/plugin-events-node';
+import express from 'express';
+import Router from 'express-promise-router';
+import { Logger } from 'winston';
+import { EventHttpRequestValidator } from './types';
+
+/**
+ * Publishes events received from their origin (e.g., webhook events from an SCM system)
+ * via HTTP endpoint and passes the request body as event payload to the registered subscribers.
+ *
+ * @public
+ */
+export class HttpEndpointEventPublisher implements EventPublisher {
+  private eventBus?: EventBus;
+
+  static fromConfig(env: {
+    config: Config;
+    logger: Logger;
+    router: express.Router;
+    validators?: Record<string, EventHttpRequestValidator>;
+  }): HttpEndpointEventPublisher {
+    const topics = env.config.getStringArray('events.modules.http.topics');
+    return new HttpEndpointEventPublisher(
+      env.logger,
+      env.router,
+      topics,
+      env.validators,
+    );
+  }
+
+  private constructor(
+    private logger: Logger,
+    router: express.Router,
+    topics: string[],
+    validators?: Record<string, EventHttpRequestValidator>,
+  ) {
+    router.use(this.createRouter(topics, validators));
+  }
+
+  async connect(eventBus: EventBus): Promise<void> {
+    this.eventBus = eventBus;
+  }
+
+  private createRouter(
+    topics: string[],
+    validators?: Record<string, EventHttpRequestValidator>,
+  ): express.Router {
+    const router = Router();
+    router.use(express.json());
+    router.use(errorHandler());
+
+    topics.forEach(topic =>
+      this.addRouteForTopic(router, topic, validators?.[topic]),
+    );
+
+    return router;
+  }
+
+  private addRouteForTopic(
+    router: express.Router,
+    topic: string,
+    validator?: EventHttpRequestValidator,
+  ): void {
+    const path = `/${topic}`;
+
+    router.post(path, async (request, response) => {
+      const rejection = await validator?.(request);
+      if (rejection) {
+        response.status(rejection.status ?? 403).json(rejection.payload ?? {});
+        return;
+      }
+
+      const eventPayload = request.body;
+      await this.eventBus!.publish(topic, eventPayload);
+
+      response.status(202).json({ status: 'accepted' });
+    });
+
+    this.logger.info(
+      `Registered /api/events/module/http${path} to receive events`,
```

Comment:
i guess technically you don't know this entire path ... but in the backend system we likely at least won't make the "events" segment configurable, so there's that

Reply:
yeah.. this was something I struggled with a bit. On one hand, we don't really know the parent path to which the router will be bound later/outside, however showing just the sub-path seemed not very helpful for readers.

Open to suggestions.

Reply:
Yeah I don't have any suggestions either. For now it's just a thing to note in the back of our heads. Maybe there will be route refs or something in the future? :P

Only thing one could do would be to find the path part of backend.baseUrl to get the `/api` bit. Let's chalk it up for the future

Reply:
I left a TODO so that we can revisit this later